mitodl/ol-django
# feat: add payment refund functionality (CyberSource)

#### Related Ticket:
https://github.com/mitodl/mitxonline/issues/571 (Part of the ticket - the refund API itself is added to this common payment_gateway app and will be consumed by the applications e.g. https://github.com/mitodl/mitxonline/pull/599)

#### Description
- Adds the configuration keys to be used for CyberSource Rest APIs
- Adds `start_refund` functionality in the Cybersource payment gateway
- Uses https://github.com/CyberSource/cybersource-rest-client-python for refunds (The SDK/Client for Cybersource REST APIs for python)

#### How should be tested:
- Acquire the keys `MITOL_PAYMENT_GATEWAY_CYBERSOURCE_MERCHANT_ID , MITOL_PAYMENT_GATEWAY_CYBERSOURCE_MERCHANT_SECRET, MITOL_PAYMENT_GATEWAY_CYBERSOURCE_MERCHANT_SECRET_KEY_ID` from CyberSource sandbox if you have any and set then in the environment variables if you are testing independently or `.env` file of the application if MITxOnline or other. (Ideally, these Keys should be from our RC's cybersource account)
- When you install this plugin you should call `start_payment` method from the CyberSource gateway with [Refund object](https://github.com/mitodl/ol-django/pull/64/files#diff-f3d33fa9650a95b4547558b3e4c6b6a0f1d0c1d062c878515ed7a160ae713fe4R76-R90). 
```
PaymentGateway.start_refund(
            ECOMMERCE_DEFAULT_PAYMENT_GATEWAY,
            refund_object,
        )
```
- The refund has the following conditions that can be tested:
    - API Success:
                Returns a ProcessorResponse object that the caller app can use to perform whatever they need
    -  API Failure ("DUPLICATE_REQUEST"):
                Raises RefundDuplicateException when the refund API call is duplicate
    -  API Failure (General):
                Raises the exception(Mostly CyberSource.rest.ApiException) to the caller to decide what their
                application behavior 

**NOTE:** 
- If you don't have CyberSource RC keys, you can use the keys from their samples test client https://github.com/CyberSource/cybersource-rest-samples-python/blob/master/data/Configuration.py#L20-L21
- REST API keys are different from Secure acceptance keys so we can't use secure acceptance keys from RC.
- For creating payments you can use https://developer.cybersource.com/api-reference-assets/index.html#payments_payments_process-a-payment. Make sure to add `true` for `capture` in the request
```
"processingInformation": {
    "capture": true
  },
```

If anything is unclear, let me know!
